casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -65,191 +66,414 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      differentPolarities(right)(base)
+      assert(!allMatch(left ++ right, base) || onePersistent(base))
+    }.pure[Effect]
+
+  }
+
+  /***
+   A send and a receive were incoming, at least one had a match, either:
+   - both were linear
+   - one was non-linear, the other had a match
+   They couldn't match the same linear event (they have different polarity)
+   They couldn't spawn more work, because either:
+   - both were linear, one of them had a match in TS
+   - one was non-linear, but the other chose to go with its match
+    */
+  object HadItsMatch extends MergeableCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      differentPolarities(left)(right)
+      crossedPolarity(left)(base)
+      crossedPolarity(right)(base)
+      assert(someMatch(left, base) || someMatch(right, base))
+    }.pure[Effect]
+
+  }
+
+  /***
+   An incoming send and an incoming receive could match each other,
+   leading to more COMMs needing to happen.
+   Mergeable if we use spatial matcher to prove they don't match.
+    */
+  object IncomingCouldMatch extends ConflictingCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      differentPolarities(left)(right)
+      assert(someMatch(left, right))
+    }.pure[Effect]
+
+  }
+
+  /***
+   There was a COMM within one of the deploys.
+   The other deploy saw none of it.
+    */
+  object VolatileEvent extends VolatileCase with MergeableCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] =
+      super.precondition(left: _*)(right: _*)(base: _*) >> {
+        allLinearWhenTwo(left)
+        allLinearWhenTwo(right)
+      }.pure[Effect]
+
+  }
+
+  /***
+   There was a COMM within one of the deploys, with one side non-linear.
+   The other deploy had an event without a match in TS, dual to the non-linear.
+   These could spawn more work.
+   Mergeable if we use spatial matcher to prove they don't match.
+    */
+  object PersistentCouldMatch extends VolatileCase with ConflictingCase {
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] =
+      super.precondition(left: _*)(right: _*)(base: _*) >> {
+        assert(onePersistent(left ++ right))
+      }.pure[Effect]
+
+  }
+
+  /***
+   There was a COMM within one of the deploys, with one side non-linear.
+   The other deploy had an event without a match in TS, of same polarity to the non-linear.
+   These could not spawn more work.
+    */
+  object PersistentNoMatch extends VolatileCase with MergeableCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] =
+      super.precondition(left: _*)(right: _*)(base: _*) >> {
```

Comment:
Can we replace the inheritance with calling a common method with a telling name?

Reply:
If you insist